File: java/src/main/java/Bots/Observation.java
package Bots;

import org.json.JSONObject;

public class Observation {

    // metadata
    int round;
    int num_rounds;
    int frame;
    // features
    int action;
    double x;
    double y;
    double heading;
    // NOTE
    // some features (named scanned_*) are only valid if the robot was scanned
    // if this flag is true, the features are valid
    boolean scanned;
    double scanned_enemy_distance;
    double scanned_enemy_bearing;

    // events
    int enemy_collisions;
    int wall_collisions;
    int shell_hits;
    int shell_wounds;
    int shell_misses;
    int shell_intercepts;

    public Observation(int round, int num_rounds, int frame, int action, double x, double y, double heading) {
        this.round = round;
        this.num_rounds = num_rounds;
        this.frame = frame;
        this.action = action;
        this.x = x;
        this.y = y;
        this.heading = heading;
    }

    public JSONObject toJson() {

        JSONObject jsn = new JSONObject();
        // metadata
        jsn.put("round", round);
        jsn.put("num_rounds", num_rounds);
        jsn.put("frame", frame);
        // features
        jsn.put("action", action);
        jsn.put("x", x);
        jsn.put("y", y);
        jsn.put("heading", heading);
        jsn.put("scanned", scanned);
        jsn.put("scanned_enemy_distance", scanned_enemy_distance);
        jsn.put("scanned_enemy_bearing", scanned_enemy_bearing);
        // events
        jsn.put("enemy_collisions", enemy_collisions);
        jsn.put("wall_collisions", wall_collisions);
        jsn.put("shell_hits", shell_hits);
        jsn.put("shell_wounds", shell_wounds);
        jsn.put("shell_misses", shell_misses);
        jsn.put("shell_intercepts", shell_intercepts);

        return jsn;
    }
}

File: python/observation.py
class Observation(object):
    pass


def json_to_observation(jsn):

    obs = Observation()

    # metadata
    obs.round = jsn['round']
    obs.num_rounds = jsn['num_rounds']
    obs.frame = jsn['frame']
    # features
    obs.action = jsn['action']
    obs.x = jsn['x']
    obs.y = jsn['y']
    obs.heading = jsn['heading']
    obs.scanned = jsn['scanned']
    obs.scanned_enemy_distance = jsn['scanned_enemy_distance']
    obs.scanned_enemy_bearing = jsn['scanned_enemy_bearing']
    # events
    obs.enemy_collisions = jsn['enemy_collisions']
    obs.wall_collisions = jsn['wall_collisions']
    obs.shell_hits = jsn['shell_hits']
    obs.shell_wounds = jsn['shell_wounds']
    obs.shell_misses = jsn['shell_misses']
    obs.shell_intercepts = jsn['shell_intercepts']

    return obs


def csv_append(filepath, obs):
    record = "{},{},{},{},{},{},{},{},{},{},{},{},{},{},{},{}".format(
        obs.round,
        obs.num_rounds,
        obs.frame,
        obs.action,
        obs.x,
        obs.y,
        obs.heading,
        obs.scanned,
        obs.scanned_enemy_distance,
        obs.scanned_enemy_bearing,
        obs.enemy_collisions,
        obs.wall_collisions,
        obs.shell_hits,
        obs.shell_wounds,
        obs.shell_misses,
        obs.shell_intercepts)

    with open(str(filepath), 'a') as handle:
        handle.write("{}\n".format(record))


def csv_create(filepath):
    header = "round,num_rounds,frame,action,x,y,"\
            "heading,scanned,scanned_enemy_distance,scanned_enemy_bearing," \
            "enemy_collisions,wall_collisions," \
            "shell_hits,shell_wounds,shell_misses,shell_intercepts"

    with open(str(filepath), 'w') as handle:
        handle.write("{}\n".format(header))
